### PR TITLE
Add user data to callback functions

### LINKED ISSUE #181: Callback functions miss user data
The callbacks `astarte_device_data_event_callback_t`, `astarte_device_unset_event_callback_t`, `astarte_device_connection_event_callback_t` and `astarte_device_disconnection_event_callback_t` are lacking a way for the user to share data with them.
This requires the callback functions implementation to rely on global variables.

A simple solution is to add an entry to the `astarte_device_config_t` struct with a pointer to user-defined data that will then also be passed to the callbacks using the `astarte_device_**_event_t` structs.

### PR DESCRIPTION
**Pull request by sorru94**
Closes #181

### CODE REVIEW

**Review by lucaato — APPROVED** [review]

File: examples/datastreams/main/src/example_task.c [hunk 1]
```diff
@@ -91,15 +91,13 @@ void astarte_example_task(void *ctx)
         ESP_LOGE(TAG, "Failed to initialize credentials");
         return;
     }
-    /*
-     * For additional ways to define the hwid of your device, see the documentation for
-     * the astarte_device_init function in astarte_device.h
-     *
-     */
```
Comment (harlem88) [inline examples/datastreams/main/src/example_task.c:98]: Why did you remove this comment?

**Review by harlem88 — CHANGES_REQUESTED** [review]
Add entry to the changelog about `user_data`.

File: examples/datastreams/main/src/example_task.c (hunk 1 above)
Comment (sorru94) [reply examples/datastreams/main/src/example_task.c:98]: Seemed very specific regarding the HWID. It also does not add useful additional information. 
It's kind of implied to check the related header for more information regarding a function.
Also, it could be better to encourage users to read the generated Doxygen instead of the header.
